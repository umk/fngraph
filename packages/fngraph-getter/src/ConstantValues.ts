import { DataRecord } from '@fngraph/data'
import {
  CONTEXT_GROUP_INVERSION,
  Falsy,
  GeneratorValue,
  Getter,
  GetterFactory,
} from '@fngraph/generator'

import { createIteration } from './IterationFunction'

type ConstantValues<V extends DataRecord> = Array<V | Falsy> | V | Falsy

export function createConstant<V extends DataRecord>(
  source: ConstantValues<V>,
): GetterFactory<never, V> {
  return function (_incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (
      _properties,
      records,
      context,
      getRecord,
      invert,
    ): AsyncGenerator<GeneratorValue> {
      const iteration = iterationF(context, invert)
      for await (const current of records) {
        context.advanceToValueGroup(current)
        const base = getRecord(current.contexts)
        if (base) {
          const valuesArr = Array.isArray(source) ? source : [source]
          if (!invert) {
            yield* iteration(current, base, valuesArr.filter(Boolean) as Array<V>)
          } else {
            if (!valuesArr.some(Boolean)) {
              yield* iteration(current, base, CONTEXT_GROUP_INVERSION as V)
            }
          }
        }
      }
    }
  }
}

export default ConstantValues
