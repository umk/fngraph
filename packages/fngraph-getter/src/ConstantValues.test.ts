import { Context, DeclarationID } from '@fngraph/data'
import {
  contextAsRecord,
  createGetContextRecord,
  createMergeContexts,
  GeneratorContext,
  GeneratorValue,
  recordAsContext,
} from '@fngraph/generator'

import { createConstant } from './ConstantValues'

describe('createConstant', () => {
  const declarationA = 'declaration A' as DeclarationID
  const declarationB = 'declaration B' as DeclarationID
  it('iterates through constant records', async () => {
    const records: AsyncGenerator<GeneratorValue> = (async function* () {
      yield { contexts: [{} as Context], n: 1 }
    })()
    const context = new GeneratorContext(0, [], false)
    const getRecord = createGetContextRecord([], createMergeContexts([]))
    const getter = createConstant([
      { [declarationA]: 1, [declarationB]: 2 },
      { [declarationA]: 3, [declarationB]: 4 },
    ])(contextAsRecord, recordAsContext)
    const derived: Array<GeneratorValue> = []
    for await (const record of getter(records, context, getRecord, false)) {
      derived.push({
        contexts: [...record.contexts],
        n: record.n,
      })
    }
    expect(derived).toMatchSnapshot()
  })
})
