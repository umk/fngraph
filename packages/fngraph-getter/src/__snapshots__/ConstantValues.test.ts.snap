// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createConstant iterates through constant records 1`] = `
[
  {
    "contexts": [
      {
        "declaration A": 1,
        "declaration B": 2,
      },
    ],
    "n": 1,
  },
  {
    "contexts": [
      {
        "declaration A": 3,
        "declaration B": 4,
      },
    ],
    "n": 0,
  },
]
`;
